src/gt4py/next/ffront/fbuiltins.py

```diff
@@ -37,103 +39,138 @@
-    )
-)
 
-astype = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[
-            ts.DeferredType(constraint=ts.FieldType),
-            ts.DeferredType(constraint=ts.FunctionType),
-        ],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=ts.FieldType),
-    )
-)
+@dataclasses.dataclass(frozen=True)
+class BuiltInFunction(Generic[R, P]):
+    name: str = dataclasses.field(init=False)
+    # `function` can be used to provide a default implementation for all `Field` implementations,
+    # e.g. a fused multiply add could have a default implementation as a*b+c, but an optimized implementation for a specific `Field`
+    function: Callable[P, R]
 
-_unary_math_builtin = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[ts.DeferredType(constraint=(ts.ScalarType, ts.FieldType))],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=(ts.ScalarType, ts.FieldType)),
-    )
-)
+    def __post_init__(self):
+        # functools.update_wrapper(
+        #     self, self.function
+        # )  # TODO figure out keeping function annotations in autocomplete
+        object.__setattr__(self, "name", f"{self.function.__module__}.{self.function.__name__}")
+
+    def __call__(self, *args: Value, **options: Any) -> Value | tuple[Value, ...]:
+        impl = self.dispatch(*args)
+        return impl(*args, **options)
+
+    def dispatch(self, *args: Value) -> Callable[P, R]:
+        arg_types = tuple(type(arg) for arg in args)
+        for atype in arg_types:
+            if (dispatcher := getattr(atype, "__gt_op_func__", None)) is not None and (
+                op_func := dispatcher(self)
+            ) is not NotImplemented:
+                return op_func
+        else:
+            return self.function
+
+    def __gt_type__(self) -> ts.FunctionType:
+        signature = inspect.signature(self.function)
+        params = signature.parameters
+
+        return ts.FunctionType(
+            pos_only_args=[
+                _type_conversion(param.annotation)
+                for param in params.values()
+                if param.kind == inspect.Parameter.POSITIONAL_ONLY
+            ],
+            pos_or_kw_args={
+                k: _type_conversion(v.annotation)
+                for k, v in params.items()
+                if v.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD
+            },
+            kw_only_args={
+                k: _type_conversion(v.annotation)
+                for k, v in params.items()
+                if v.kind == inspect.Parameter.KEYWORD_ONLY
+            },
+            returns=_type_conversion(signature.return_annotation),
+        )
+
+
+def builtin_function(fun: Callable[P, R]) -> BuiltInFunction[R, P]:
+    return BuiltInFunction(fun)
+
+
+@builtin_function
+def neighbor_sum(
+    field: Field,
+    /,
+    axis: Dimension,
+) -> Field:
+    raise NotImplementedError()
```

Comment:
Till: could we have (parts of) the relevant logic in this function, such that debugging (inspecting) is direct.